src/physics/base/PhysicsTrackView.i.hh

```diff
@@ -200,6 +201,94 @@ CELER_FUNCTION auto PhysicsTrackView::value_grid(ValueGridType     table_type,
+ * \f[
+     p = 1 - \exp \left( -\int_{E_0}^{E_1} n \sigma(E) \dif s \right),
+ * \f]
+ * where \f$ E_0 \f$ is the pre-step energy, \f$ E_1 \f$ is the post-step
+ * energy, \em n is the atom density, and \em s is the interaction length.
+ *
+ * At the start of the step, the maximum value of the cross section over the
+ * step \f$ \sigma_{max} \f$ is estimated and used as the macroscopic cross
+ * section for the process rather than \f$ \sigma_{E_0} \f$. After the step,
+ * the new value of the cross section \f$ \sigma(E_1) \f$ is calculated, and
+ * the discrete interaction for the process occurs with probability
+ * \f[
+     p = \frac{\sigma(E_1)}{\sigma_{\max}}.
+ * \f]
+ *
+ * See section 7.4 of the Geant4 Physics Reference (release 10.6) for details.
+ */
+CELER_FUNCTION bool PhysicsTrackView::use_integral(ParticleProcessId ppid) const
+{
+    CELER_EXPECT(ppid < this->num_particle_processes());
+    return this->energy_max(ppid) > 0;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Energy corresponding to the maximum cross section for the material.
+ *
+ * If for this particle process the macro_xs \c ValueTable is "true" and \c
+ * energy_max is assigned, the process also has energy loss tables and the
+ * integral approach is being used. If \c energy_max[material] is nonzero,
+ * those tables are present for this material.
+ */
+CELER_FUNCTION real_type PhysicsTrackView::energy_max(ParticleProcessId ppid) const
+{
+    CELER_EXPECT(ppid < this->num_particle_processes());
+    ValueTableId table_id
+        = this->process_group().tables[ValueGridType::macro_xs][ppid.get()];
+
+    CELER_ASSERT(table_id);
+    const ValueTable& table = params_.value_tables[table_id];
+
+    real_type result = 0;
+    if (table && !table.energy_max.empty())
+    {
+        CELER_ASSERT(material_ < table.energy_max.size());
+        result = params_.reals[table.energy_max[material_.get()]];
+    }
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate macroscopic cross section for the process.
+ *
+ * If this is an energy loss process, this returns the estimate of the maximum
+ * cross section over the step. If the energy of the global maximum of the
+ * cross section (calculated at initialization) is in the interval \f$ [\xi
+ * E_0, E_0) \f$, where \f$ \E_0 \f$ is the pre-step energy and \f$ \xi \f$ is
+ * \c energy_fraction, \f$ \sigma_{\max} \f$ is set to the global maximum.
+ * Otherwise, \f$ \sigma_{\max} = \max( \sigma(E_0), \sigma(\xi E_0) ) \f$.
+ */
+CELER_FUNCTION real_type PhysicsTrackView::calc_xs(ParticleProcessId ppid,
+                                                   ValueGridId       grid_id,
+                                                   MevEnergy energy) const
+{
+    auto      calc_xs    = this->make_calculator<XsCalculator>(grid_id);
+    real_type energy_max = this->energy_max(ppid);
+    if (energy_max > 0)
```

Comment:
Eh, I didn't love this either. This is essentially the same as calling `use_integral(ppid)`, which evaluates to true if `energy_max > 0` for the given material/ppid. Since we need `energy_max` in this function I just accessed and evaluated it directly, but I'm sure there's a better/clearer way of doing this. I'll add some documentation at least.